### PR TITLE
SASS-8635 P&L:Loss brought forward - create taxYearBroughtForwardFrom…

### PR DESCRIPTION
**Pull request by kuldeepwadhwa9**
… content page 3

### CODE REVIEW

**Review by Dweik97 — APPROVED** [review]

File: app/models/journeys/adjustments/WhichYearIsLossReported.scala [hunk 1]
```diff
@@ -21,16 +21,24 @@ import play.api.i18n.Messages
...
 
+import java.time.Year
+
 sealed trait WhichYearIsLossReported
 
 object WhichYearIsLossReported extends Enumerable.Implicits {
 
-  case object Year2018to2019 extends WithName("2018to2019") with WhichYearIsLossReported
-  case object Year2019to2020 extends WithName("2019to2020") with WhichYearIsLossReported
+  case object currentTaxYearMinusFour  extends WithName(generateYears().last) with WhichYearIsLossReported
```
Comment (lukastymo) [inline app/models/journeys/adjustments/WhichYearIsLossReported.scala:30]: we changed hardcoded 2018to2019 with generateYears() - but it is still hardcoded. The application need to be restarted if we want to support more years. I wouldn't use generateYears as enum.

We have two choices:
- We hardcode like it was but for more years, and each year we remove the oldest one and add newer one. Not perfect, but easy to implement.
- We move away from enum as the value is dynamic always 5 tax years.

File: app/models/journeys/adjustments/WhichYearIsLossReported.scala [hunk 2]
```diff
@@ -44,4 +52,14 @@ object WhichYearIsLossReported extends Enumerable.Implicits {
   implicit val enumerable: Enumerable[WhichYearIsLossReported] =
     Enumerable(values.map(v => v.toString -> v): _*)
 
+  private def generateYears(upTo: Int = 5): Seq[String] = {
```
Comment (lukastymo) [inline app/models/journeys/adjustments/WhichYearIsLossReported.scala:55]: I would probably use hardcoded enums, and update it every year. But if it is not doable I would implement dynamic function and put it in our common place where we have year-dependent code.

This function will probably go, but just for future: this function is pretty complex, I would unit test it, especially edge cases. To do this you need to make `private[adjustments]` so it is open in the same package and easily tested

**Review by lukastymo — CHANGES_REQUESTED** [review]
I'm not convince 5 dynamic tax years should be implemented as enum

**Review by lukastymo — APPROVED** [review]
